File: web/src/api/OrganizationAuthApi.ts
import Api, { apiState } from 'api/Api';
import { selectorFamily } from 'recoil';
import OrganizationAuthRep from 'rep/OrganizationAuthRep';

class OrganizationAuthApi {
  private readonly api: Api;

  constructor(api: Api) {
    this.api = api;
  }

  async get(organizationId: string): Promise<OrganizationAuthRep | null> {
    return await this.api.request<OrganizationAuthRep | null>({
      method: 'GET',
      path: `/organizations/${organizationId}/auths`,
    });
  }

  async getByHostname(hostname: string): Promise<OrganizationAuthRep | null> {
    return await this.api.request<OrganizationAuthRep | null>({
      method: 'GET',
      path: '/organization-auths',
      qp: new URLSearchParams({ hostname }),
    });
  }

  async create(organizationId: string, creator: OrganizationAuthRep.Creator): Promise<OrganizationAuthRep> {
    return await this.api.request<OrganizationAuthRep, OrganizationAuthRep.Creator>({
      method: 'POST',
      path: `/organizations/${organizationId}/auths`,
      body: creator,
    });
  }

  async update(authId: string, updater: OrganizationAuthRep.Updater): Promise<OrganizationAuthRep> {
    return await this.api.request<OrganizationAuthRep, OrganizationAuthRep.Updater>({
      method: 'PATCH',
      path: `/organization-auths/${authId}`,
      body: updater,
    });
  }

  async delete(authId: string): Promise<OrganizationAuthRep> {
    return await this.api.request<OrganizationAuthRep>({
      method: 'DELETE',
      path: `/organization-auths/${authId}`,
    });
  }
}

export const organizationAuthApiState = selectorFamily<OrganizationAuthApi, { authenticated: boolean; }>({
  key: 'api/organizationAuth',
  get: ({ authenticated }) => ({ get }) => {
    const api = get(apiState({ authenticated }));
    return new OrganizationAuthApi(api);
  },
});

File: feature/organization/rest-interface/src/main/kotlin/limber/api/organizationAuth/OrganizationAuthApi.kt
package limber.api.organizationAuth

import io.ktor.http.HttpMethod
import jakarta.validation.Valid
import limber.feature.rest.QueryParam
import limber.feature.rest.RestEndpoint
import limber.rep.organizationAuth.OrganizationAuthRep

public object OrganizationAuthApi {
  public data class Get(
    val authId: String,
  ) : RestEndpoint<Nothing>() {
    override val method: HttpMethod = HttpMethod.Get
    override val path: String = "/organization-auths/$authId"
  }

  public data class GetByOrganization(
    val organizationId: String,
  ) : RestEndpoint<Nothing>() {
    override val method: HttpMethod = HttpMethod.Get
    override val path: String = "/organizations/$organizationId/auths"
  }

  public data class GetByHostname(
    val hostname: String,
  ) : RestEndpoint<Nothing>() {
    override val method: HttpMethod = HttpMethod.Get
    override val path: String = "/organization-auths"
    override val qp: List<QueryParam> = listOf(::hostname)
  }

  public data class Create(
    val organizationId: String,
    @Valid override val body: OrganizationAuthRep.Creator?,
  ) : RestEndpoint<OrganizationAuthRep.Creator>() {
    override val method: HttpMethod = HttpMethod.Post
    override val path: String = "/organizations/$organizationId/auths"
  }

  public data class Update(
    val authId: String,
    @Valid override val body: OrganizationAuthRep.Update?,
  ) : RestEndpoint<OrganizationAuthRep.Update>() {
    override val method: HttpMethod = HttpMethod.Patch
    override val path: String = "/organization-auths/$authId"
  }

  public data class Delete(
    val authId: String,
  ) : RestEndpoint<Nothing>() {
    override val method: HttpMethod = HttpMethod.Delete
    override val path: String = "/organization-auths/$authId"
  }
}
